n3fit/src/n3fit/hyper_optimization/mongofiletrials.py

```diff
@@ -264,21 +266,28 @@ def start_mongo_workers(
                     # avoid memory fragmentation issues?
                     # my_env["TF_GPU_ALLOCATOR"] = "cuda_malloc_async"
 
+                # create log files to redirect the mongo-workers output
+                mongo_workers_logfile = f"mongo-worker_{i+1}_{self.output_path}.log"
+                log_file = open(mongo_workers_logfile, 'w')
```

Comment:
Can't you use a context manager with `with` here?

Reply:
Thanks @Radonirinaunimi for the very helpful comments. The reason is that these files cannot be closed until each mongo worker is done with its job. They are close here:
https://github.com/NNPDF/nnpdf/blob/4974bda5ec12dce1535b017fe6e06a993b2407b8/n3fit/src/n3fit/hyper_optimization/mongofiletrials.py#L288-L290
after the mongo workers termininate.

Reply:
Yes, I indeed noticed that. And the reason why I was asking about the context manager was because I am a bit worried about the fact these files are only closed when the workers stopped. Since you are only flushing stdout, surely these can be appended.

Reply:
@Radonirinaunimi and @APJansen. It seems that the use of `with` while opening the mongo worker file just work. Although they are closed after the execution of the `start_mongo_workers` method, they are inherently tied to the subprocess and can be written as expected; see [here](https://github.com/NNPDF/nnpdf/blob/4bdcca4d7c04ef7140cdfc1178132a04f8e91845/n3fit/src/n3fit/hyper_optimization/mongofiletrials.py#L273). 

The only thing is that I would prefer opening them with `write` mode only as I foresee unexpected issues when using the `append` mode: *e.g.*: if you have the mongo logs in the same directory and want to rerun a previous calculation that gone wrong for any reason, the old files get appended with the new results. The drawback however it that if you want to restart your calculation, your mongo logs will be reset and only have info from the current run (not from the previous ones). But I think this is OK since all the info we need to make any post-processing of the data are contained in the `tries.json` and `hyperopt-db.tar.gz` files.

Reply:
Thanks a lot @Cmurilochem! This is now much cleaner and (I'd argue) safer.

> The only thing is that I would prefer opening them with write mode only as I foresee unexpected issues when using the append mode: e.g.: if you have the mongo logs in the same directory and want to rerun a previous calculation that gone wrong for any reason, the old files get appended with the new results.

Yes, this is exactly the intended behavior and the fact that the logs are lost once restarting shouldn't be an issue at all (they can be copied elsewhere in-between) and as you said all the main information we need are saved correctly.